File: JS_OC_Swift/JS_OC/JS_JavaScriptCore/JS_JavaScriptCore/ViewController.m
//
//  ViewController.m
//  JS_JavaScriptCore
//
//  Created by xinyu on 2018/3/26.
//  Copyright © 2018年 MaChat. All rights reserved.
//

#import "ViewController.h"


@interface ViewController ()<UIWebViewDelegate,JSDelegate>
@property (nonatomic, strong) UIWebView *webView;
@property (nonatomic, strong) JSContext *context;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.webView = [[UIWebView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height)];
    [self.view addSubview:self.webView];
    
    self.webView.delegate = self;
    // UIWebView 滚动的比较慢，这里设置为正常速度
    self.webView.scrollView.decelerationRate = UIScrollViewDecelerationRateNormal;
    
    NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL fileURLWithPath:[[NSBundle mainBundle] pathForResource:@"index" ofType:@"html"]]];
    [self.webView loadRequest:request];
}

#pragma mark - UIWebViewDelegate
- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType {
    return YES;
}

- (void)webViewDidFinishLoad:(UIWebView *)webView {
    JSContext *context = [self.webView valueForKeyPath:@"documentView.webView.mainFrame.javaScriptContext"];
    self.context = context;
    //设置异常处理
    self.context.exceptionHandler = ^(JSContext *context, JSValue *exception) {
        [JSContext currentContext].exception = exception;
        NSLog(@"exception:%@",exception);
    };
    //注入JS需要的“OC”对象,该对象与html中的保持一致即可
    self.context[@"OC"] = self;
    // JS 调用OC 方法
    [self showMessage:context];
    [self showTitleAndMessage:context];
    [self doSomethingThenCallBack:context];
}

#pragma mark - Methods
- (void)showMessage:(JSContext *)context {
    __weak typeof(self) weakSelf = self;
    context[@"showMessage"] = ^(NSString *message) {
        UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:@"提示" message:message preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil];
        [alertCtr addAction:cancel];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf presentViewController:alertCtr animated:YES completion:nil];
        });
    };
}

- (void)showTitleAndMessage:(JSContext *)context {
    __weak typeof(self) weakSelf = self;
    context[@"showTitleAndMessage"] = ^(NSString *title, NSString *message){
        UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil];
        [alertCtr addAction:cancel];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf presentViewController:alertCtr animated:YES completion:nil];
        });
    };
}

- (void)doSomethingThenCallBack:(JSContext *)context {
    __weak typeof(self) weakSelf = self;
    self.context[@"doSomethingThenCallBack"] = ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:@"提示" message:nil preferredStyle:UIAlertControllerStyleAlert];
            [alertCtr addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
                textField.placeholder = @"请输入传入的数据!";
                [textField addTarget:weakSelf action:@selector(textFieldChange:) forControlEvents:UIControlEventEditingChanged];
            }];
            UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:nil];
            [alertCtr addAction:cancel];
            UIAlertAction *ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                UITextField *textField = [alertCtr.textFields firstObject];
                //OC 调用 JS 方法
                
                //方法一
                JSValue *callback = weakSelf.context[@"callback"];
                [callback callWithArguments:@[textField.text]];
                
            }];
            [alertCtr addAction:ok];
            ok.enabled = NO;
            
            [weakSelf presentViewController:alertCtr animated:YES completion:nil];
        });
    };
}

#pragma mark - JSDelegate
- (void)showMessageToYou:(NSString *)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:@"问候" message:message preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil];
        [alertCtr addAction:cancel];
        [self presentViewController:alertCtr animated:YES completion:nil];
    });
}

- (void)doActionCallBack {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alertCtr = [UIAlertController alertControllerWithTitle:@"提示" message:nil preferredStyle:UIAlertControllerStyleAlert];
        [alertCtr addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
            textField.placeholder = @"请输入传入的数据!";
            [textField addTarget:self action:@selector(textFieldChange:) forControlEvents:UIControlEventEditingChanged];
        }];
        UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:nil];
        [alertCtr addAction:cancel];
        UIAlertAction *ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            UITextField *textField = [alertCtr.textFields firstObject];
            //方法二
            NSString *jsStr = [NSString stringWithFormat:@"callback('%@')",textField.text];
            [weakSelf.context evaluateScript:jsStr];
        }];
        [alertCtr addAction:ok];
        ok.enabled = NO;
        
        [weakSelf presentViewController:alertCtr animated:YES completion:nil];
    });
}

- (void)showA:(NSString *)aString andB:(NSString*)bString {
    JSValue *alertCallback = self.context[@"alertCallback"];
    [alertCallback callWithArguments:@[aString,bString]];
}

#pragma mark - UITextField
- (void)textFieldChange:(UITextField *)sender {
    UIAlertController *alertCtr = (UIAlertController *)self.presentedViewController;
    UIAlertAction *okAction = [alertCtr.actions lastObject];
    if (sender.text.length > 0) {
        okAction.enabled = YES;
    } else {
        okAction.enabled = NO;
    }
}

@end

File: JS_Swift/JS_JavaScriptCore/JS_JavaScriptCore/ViewController.swift
//
//  ViewController.swift
//  JS_JavaScriptCore
//
//  Created by xinyu on 2018/3/27.
//  Copyright © 2018年 MaChat. All rights reserved.
//

import UIKit
import JavaScriptCore

@objc protocol JSDelegate :JSExport {
    //包含参数的func,需要注意参数名对函数名称的影响
    func showMessageToYou(_ message:String)
    
    /*
     对应html中“showAAndB”,此方法包含两个参数，需要在参数前加“_”
     func showA(_ aString: String, andB: String)
     func showAAndB(_ aString:String,_ bStr:String)
     以上两个方法等同
     */
    func showAAndB(_ aString:String,_ bStr:String)
    
    func doActionCallBack()
}

class ViewController: UIViewController,UIWebViewDelegate,JSDelegate {

    let webView = UIWebView.init(frame: .zero)
    var context : JSContext?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        webView.frame = CGRect.init(x: 0, y: 0, width: self.view.frame.width, height: self.view.frame.height)
        view.addSubview(self.webView)
        webView.delegate = self
        webView.scrollView.decelerationRate = UIScrollViewDecelerationRateNormal
        let request = NSURLRequest(url:URL(fileURLWithPath: Bundle.main.path(forResource: "index", ofType: "html")!))
        webView.loadRequest(request as URLRequest)
        webView.backgroundColor = UIColor.black
    }
    //MARK: - UIWebViewDelegate
    func webView(_ webView: UIWebView, shouldStartLoadWith request: URLRequest, navigationType: UIWebViewNavigationType) -> Bool {
        return true
    }
    
    func webViewDidFinishLoad(_ webView: UIWebView) {
        context = webView.value(forKeyPath: "documentView.webView.mainFrame.javaScriptContext") as? JSContext
        //设置异常处理
        context?.exceptionHandler = { (context, exception) in
            print("exception：", exception as Any)
        }
        //注入JS需要的“OC”对象,该对象与html中的保持一致即可
        context?.setObject(self, forKeyedSubscript:  "OC" as NSCopying & NSObjectProtocol)
    }
    
    //MARK: - JSDelegate
    func showMessageToYou(_ message: String) {
        showAlertControlelr(title: "问候", message: message)
    }
    
    func showAAndB(_ aString: String, _ bStr: String) {
        //OC 调用 JS
        let callBack = context?.objectForKeyedSubscript("alertCallback")
        callBack?.call(withArguments: [aString,bStr])
    }
 
    func doActionCallBack() {        
        DispatchQueue.main.async {
            let alert = UIAlertController(title: "提示", message: nil, preferredStyle: .alert)
            alert.addTextField { (textField) in
                textField.placeholder = "请输入传送的数据"
                textField.addTarget(self, action: #selector(self.textFieldChange(textField:)), for:.editingChanged)
            }
            let cancel = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
            alert.addAction(cancel)
            let okAction = UIAlertAction(title: "OK", style: .default) { (action) in
                let textField = alert.textFields?.first
                //OC 调用JS
                let jsStr = String(format:"callback('%@')",(textField?.text)!)
                self.context?.evaluateScript(jsStr)
                
            }
            alert.addAction(okAction)
            okAction.isEnabled = false
            self.present(alert, animated: true, completion: nil)
        }
    }
    //MARK: - 
    func showAlertControlelr(title:String , message:String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let cancel = UIAlertAction(title: "OK", style: .cancel, handler: nil)
        alert.addAction(cancel)
        present(alert, animated: true, completion: nil)
    }
    
    @objc func textFieldChange(textField:UITextField) {
        let alert = presentedViewController as! UIAlertController
        let okAction = alert.actions.last
        let count = textField.text!.count 
        if (count > 0) {
            okAction?.isEnabled = true
        } else {
            okAction?.isEnabled = false
        }
    }
}

